map/extrapolation/extrapolator.cpp

```diff
@@ -0,0 +1,172 @@
+#include "map/extrapolation/extrapolator.hpp"
+
+#include "geometry/distance_on_sphere.hpp"
+
+#include "platform/platform.hpp"
+
+#include "base/logging.hpp"
+
+#include <chrono>
+#include <memory>
+
+namespace
+{
+uint64_t constexpr kMaxExtrapolationTimeMs = 1000;
+uint64_t constexpr kExtrapolationPeriodMs = 200;
+double constexpr kMaxExtrapolationSpeedMPS = 120.0;
+
+double Delta(double param1, double param2, uint64_t timeBetweenPointsMs, uint64_t timeAfterParam2Ms)
+{
+  return (param2 - param1) * timeAfterParam2Ms / timeBetweenPointsMs;
+}
+
+class LinearExtrapolator
+{
+public:
+  LinearExtrapolator(uint64_t timeBetweenMs, uint64_t timeAfterMs)
+    : m_timeBetweenMs(timeBetweenMs), m_timeAfterMs(timeAfterMs)
+  {
+  }
+
+  double Extrapolate(double x1, double x2)
+  {
+    return x2 + (x2 - x1) / m_timeBetweenMs * m_timeAfterMs;
+  }
+
+private:
+  uint64_t m_timeBetweenMs;
+  uint64_t m_timeAfterMs;
+};
+}  // namespace
+
+namespace extrapolation
+{
+using namespace std;
+
+location::GpsInfo LinearExtrapolation(location::GpsInfo const & gpsInfo1,
+                                      location::GpsInfo const & gpsInfo2,
+                                      uint64_t timeAfterPoint2Ms)
+{
+  auto const timeBetweenPointsMs =
+      static_cast<uint64_t>((gpsInfo2.m_timestamp - gpsInfo1.m_timestamp) * 1000);
+
+  location::GpsInfo extrapolated = gpsInfo2;
+  LinearExtrapolator extrapolator(timeBetweenPointsMs, timeAfterPoint2Ms);
+
+  extrapolated.m_timestamp += timeAfterPoint2Ms;
+
+  extrapolated.m_longitude = extrapolator.Extrapolate(gpsInfo1.m_longitude, gpsInfo2.m_longitude);
+
+  extrapolated.m_latitude = extrapolator.Extrapolate(gpsInfo1.m_latitude, gpsInfo2.m_latitude);
+
+  extrapolated.m_horizontalAccuracy =
+      extrapolator.Extrapolate(gpsInfo1.m_horizontalAccuracy, gpsInfo2.m_horizontalAccuracy);
+
+  extrapolated.m_altitude = extrapolator.Extrapolate(gpsInfo1.m_altitude, gpsInfo2.m_altitude);
+
+  if (gpsInfo1.HasVerticalAccuracy() && gpsInfo2.HasVerticalAccuracy())
+  {
+    extrapolated.m_verticalAccuracy =
+        extrapolator.Extrapolate(gpsInfo1.m_verticalAccuracy, gpsInfo2.m_verticalAccuracy);
+  }
+
+  if (gpsInfo1.HasBearing() && gpsInfo2.HasBearing())
```

Comment:
So, when old bearing was about 10 degrees, and new bearing is about 350 degrees, extrapolation is 180 degrees? This is wrong. Please, don't use such methods for periodic random variables.

Reply:
Now I'm removing the if statement and the next line. That means: `result.m_bearing = gpsInfo2.m_bearing;`. And I'm putting a todo that another way of founding an average value should be considered.